### ISSUE TITLE
Discrepancy in Results and Evaluation Compared to Paper

### ISSUE THREAD

**Issue by uchon2287883** [NazirNayal8/RbA#12]
Our goal was to reproduce the experiment from Chapter 4 of the paper (https://arxiv.org/pdf/2211.14293.pdf) and verify if we could achieve similar results in evaluation as in the paper or in the downloadable model on GitHub (Cityscapes Inlier Training Swin-B (1 dec layer) (https://github.com/NazirNayal8/RbA/blob/main/MODEL_ZOO.md)).
Unfortunately, we found that we achieved significantly worse results than expected. I would like to find out if I made any mistakes during training and evaluation or why my results deviate so much from those in the paper.

First, I read the readme file in the repository (https://github.com/NazirNayal8/RbA/tree/main). The installation went smoothly, although I should note that I did not create a separate Conda environment.
In preparing the datasets for training RbA (https://github.com/NazirNayal8/RbA/blob/main/datasets/README.md), I followed the instructions exactly, except that I created empty folders for mapillary_vistas and coco since I only need the Cityscapes dataset.
We only trained the Swin-B model (1 dec layer) as specified in the configuration file on GitHub (https://github.com/NazirNayal8/RbA/blob/main/MODEL_ZOO.md), using the pretrained models for initialization.

Training was done on two NVIDIA GeForce RTX 3090 GPUs, each with 24 GB of memory.

`python train_net.py \
--config-file configs/cityscapes/semantic-segmentation/swin/single_decoder_layer/maskformer2_swin_base_IN21k_384_bs16_90k_1dl.yaml \
--num-gpus 2 \
OUTPUT_DIR model_logs/swin_b_1dl/ `

We only changed two lines in the repository: first, we changed "IMS_PER_BATCH:" from 16 to 8 in the Base-Cityscapes-SemanticSegmentation.yaml due to hardware issues. Second, we commented out "#bdd100k=BDD100KSeg(hparams=bdd100k_config, mode='val', transforms=transform, image_size=(720, 1280))," in support.py.
Evaluation was performed as follows:

`python3 evaluate_ood.py --out_path results_test/ --models_folder ckpts --datasets_folder datasets --model_mode all --dataset_mode all --store_anomaly_scores`

During evaluation, there is definitely no error. 
If you download the ckpt model (https://drive.google.com/file/d/13IJs_Kk1PMBVVxCN90HZZuuV1YcWZ0am/view?usp=sharing)  from Github and evaluate it you get the same results as the paper. So something must have gone wrong during training. Do you think it's because I reduced the batch size from 16 to 8?
Some Pictures of the Evaluation: 

The first three pictures shows the results of the model downloaded from the repository: 

![image](https://github.com/NazirNayal8/RbA/assets/90455844/1baf4277-bf9d-42a8-9f1c-9c5ff87d25e2)
![image](https://github.com/NazirNayal8/RbA/assets/90455844/ce52608c-e2f8-4b67-be0f-fb55b78d7d39)
![image](https://github.com/NazirNayal8/RbA/assets/90455844/f4a136a3-9b7d-4ee9-b637-c28a5a4542dc)

These three pictures shows the results from the model we trained:

![image](https://github.com/NazirNayal8/RbA/assets/90455844/126032c4-6aef-4162-b9de-8f2d28bd87ef)
![image](https://github.com/NazirNayal8/RbA/assets/90455844/6dcedd0b-24e3-4f43-be49-f44707cf663b)
![image](https://github.com/NazirNayal8/RbA/assets/90455844/5e1c6e5e-6a66-4257-93df-dfb2349affed)

We can see that our model does not recognize all anomalies or mask them as well as the model downloaded from the repository.

**Comment by Orange-rightperson** [comment]
I met the same problem

**Comment by IrammHamdard** [comment]
I tried training more than 90000 iteration and also with smaller learning rate but did not get any better results. 
Also i trained with gradient accumulation and did not get any better results. the fpr95 is to high. 
Here some results training with gradient accumulation every second step and training with batch size 8. 
![image](/uploads/62c748645542847e9da3a59da2fe000b/image.png)
![IoU_nIOU_gradient](/uploads/d5eda26bc0d5433955004b9658e393fb/IoU_nIOU_gradient.png)
![Screenshot_from_2024-05-15_16-49-41](/uploads/0111c8db9cef2fce5b7ea8a731b33802/Screenshot_from_2024-05-15_16-49-41.png)
![image](/uploads/c37e0d90d0430d90a6fcee3ec7326188/image.png)

**Comment by qiqi-x** [comment]
Hello, I encountered the same issue as described. I am using two NVIDIA GeForce RTX 3090 GPUs, each with 24 GB of memory, and have set IMS_PER_BATCH to 8. Following your advice, I trained for 180K iterations, and the results are as follows: 
![image](https://github.com/user-attachments/assets/4c69dd07-0e99-4f18-8d5b-56164e9b6a9b)
The data does not match the numbers in your paper. Could you please tell me what might be the issue here?

**Comment by NazirNayal8** [comment]
Greetings @qiqi-x , which model version are you using? And is this with or without outlier supervision? If you provide this info along with the config file you used we might be able to help.

**Comment by qiqi-x** [comment]
Hello, thank you very much for your reply. I am using the Cityscapes Inlier Training Swin-B (1 dec layer) model, and the configuration file I used is as follows:https://github.com/NazirNayal8/RbA/blob/main/configs/cityscapes/semantic-segmentation/swin/single_decoder_layer/maskformer2_swin_base_IN21k_384_bs16_90k_1dl.yaml.I only performed inlier supervision and did not use outlier supervision.Now I don't know where the problem lies.
